### PR TITLE
Use MoreElements to collect methods

### PR DESCRIPTION
**Pull request by pkoenig10**
We use this utility in most of our other annotation processors that generate delegating implementations, including the conjure undertow annotation processor and a number of our internal annotation processors.

This simplifies the code and will make it easier to support selective per-method instrumentation in a future PR.

The one caveat is that this annotation processor no longer handles that case when methods with the same signature are inherited from difference ancestors. This seems acceptable since 1) this case is likely rare and 2) this isn't supported by any of our other annotation processors that use `MoreElements`. If this becomes an issue we add some logic to handle this case.

```java
interface A {
    void foo();
}

interface B {
    void foo();
}

@Instrument
interface MyInterface extends A, B {}
```

### CODE REVIEW

File: tritium-processor/src/main/java/com/palantir/tritium/processor/TritiumAnnotationProcessor.java
Comment (pkoenig10) [inline tritium-processor/src/main/java/com/palantir/tritium/processor/TritiumAnnotationProcessor.java:177]: Is there a particular reason why we need to handle this case here but not in our other delegate-generating annotation processors that use `MoreElements`?

File: tritium-processor/src/test/java/com/palantir/tritium/processor/TritiumProcessorTest.java
Comment (pkoenig10) [inline tritium-processor/src/test/java/com/palantir/tritium/processor/TritiumProcessorTest.java:116]: While unusual, there's no reason why we can't support empty interfaces. It was a trivial fix, so I've fixed it here.

**Review by schlosna — APPROVED** [review]
Seems reasonable.

What is the behavior now if method is inherited from two super types? Will we fail at compile or runtime?

**Comment by pkoenig10** [comment]
It will fail at compile time, because we will generate the method multiple times.

```
Compilation produced the following diagnostics:
/SOURCE_OUTPUT/com/palantir/tritium/examples/InstrumentedMyInterface.java:57: error: method foo() is already defined in class com.palantir.tritium.examples.InstrumentedMyInterface
    public void foo() {
                ^
Generated Source Files
======================

/SOURCE_OUTPUT/com/palantir/tritium/examples/InstrumentedMyInterface.java:
package com.palantir.tritium.examples;

import com.palantir.tritium.annotations.internal.InstrumentationBuilder;
import com.palantir.tritium.api.event.InstrumentationFilter;
import com.palantir.tritium.event.Handlers;
import com.palantir.tritium.event.InvocationContext;
import com.palantir.tritium.event.InvocationEventHandler;
import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
import java.lang.reflect.Method;
import javax.annotation.processing.Generated;

@Generated("com.palantir.tritium.processor.TritiumAnnotationProcessor")
public final class InstrumentedMyInterface implements MyInterface {
    private static final Method FOO;

    private static final Method FOO1;

    static {
        try {
            FOO = MyInterface.class.getMethod("foo");
            FOO1 = MyInterface.class.getMethod("foo");
        } catch (NoSuchMethodException e) {
            throw new RuntimeException(e);
        }
    }

    private final MyInterface delegate;

    private final InvocationEventHandler<InvocationContext> handler;

    private final InstrumentationFilter filter;

    private InstrumentedMyInterface(
            MyInterface delegate, InvocationEventHandler<InvocationContext> handler, InstrumentationFilter filter) {
        this.delegate = delegate;
        this.handler = handler;
        this.filter = filter;
    }

    @Override
    public void foo() {
        if (this.handler.isEnabled()) {
            InvocationContext invocationContext = Handlers.pre(this.handler, this.filter, this, FOO, new Object[] {});
            try {
                this.delegate.foo();
                Handlers.onSuccess(this.handler, invocationContext);
            } catch (Throwable throwable) {
                Handlers.onFailure(this.handler, invocationContext, throwable);
                throw throwable;
            }
        } else {
            this.delegate.foo();
        }
    }

    @Override
    public void foo() {
        if (this.handler.isEnabled()) {
            InvocationContext invocationContext = Handlers.pre(this.handler, this.filter, this, FOO1, new Object[] {});
            try {
                this.delegate.foo();
                Handlers.onSuccess(this.handler, invocationContext);
            } catch (Throwable throwable) {
                Handlers.onFailure(this.handler, invocationContext, throwable);
                throw throwable;
            }
        } else {
            this.delegate.foo();
        }
    }

    @Override
    public String toString() {
        return "InstrumentedMyInterface{" + delegate + '}';
    }

    public static InstrumentationBuilder<MyInterface, MyInterface> builder(MyInterface delegate) {
        return new InstrumentationBuilder<MyInterface, MyInterface>(
                MyInterface.class, delegate, InstrumentedMyInterface::new);
    }

    public static MyInterface trace(MyInterface delegate) {
        return InstrumentationBuilder.trace(MyInterface.class, delegate, InstrumentedMyInterface::new);
    }

    public static MyInterface instrument(MyInterface delegate, TaggedMetricRegistry registry) {
        return builder(delegate).withTaggedMetrics(registry).withTracing().build();
    }
}
```

**Comment by schlosna** [comment]
> It will fail at compile time, because we will generate the method multiple times

If this fails fast at compile time, that seems fine to release and roll out. We'll have to keep an eye on any repos that fail to bump tritium or server framework (the latter being more problematic where we would need to add support for these or revert).

**Comment by schlosna** [comment]
@pkoenig10 defer to you when you want to merge & release.

**Comment by pkoenig10** [comment]
I'd like to get a +1 from @carterkozak before merging this, particularly with regards to https://github.com/palantir/tritium/pull/1747#discussion_r1249745539.

**Comment by pkoenig10** [comment]
For more context, my plan for per-method instrumentation is to only instrument methods that are declared directly in the wrapped interface.

This nicely solves the problem of multiple inherited methods with the same signature. It also requires users to be explicit about which methods they want to instrument, which feels like a desirable property.

I'd probably go so far as to say that this should be the recommended way for users to instrument their interfaces and that we should soft-deprecate the class level `@Instrument` annotation.

```java
interface A {
    void foo();
}

interface B {
    void foo();
    void bar();
}

// Only foo will be instrumented
interface MyInterface extends A, B {
    @Instrument
    @Override
    void foo();
}
```

**Comment by carterkozak** [comment]
@pkoenig10 and I chatted offline -- current plan is to try moving the tritium processor to our delegate-processors library. I'll open-source delegate-processors if that goes smoothly. This way we get all the delegate implementation factory pieces in one place, and don't have to worry about breaking edge cases where an implementation inherits from multiple interfaces.

**Comment by pkoenig10** [comment]
Closing in favor of https://github.com/palantir/tritium/pull/1766

**Pull request closed without merging**
